File: Matlab/Measurment_Processing/mgen.m
%m序列发生器
function [out,pState] = mgen(g,state,N)
%输入 g: m序列生成多项式 (10进制输入）
%  state: 寄存器初始状态 （10进制输入）
%     N: 输出序列长度
% test g =11; state=3; N=15;
gen = dec2bin(g) - 48;
M = length(gen);                
curState = dec2bin(state,M-1) - 48;

for k=1:N
    out(k) = curState(M-1);
    a = rem( sum( gen(2:end).*curState ),2 );
    curState = [a curState(1:M-2)];
end
pState = bin2dec( char(curState+48) );

File: Python/mgen.py
#!/usr/bin/env python
# encoding: utf-8

import numpy as np

def mgen(g, state, N):
    gen = list(bin(g)[2::])
    gen = chartoint(gen)
    gen = np.array(gen)
    M = len(gen)
    cur_state = list(bin(state)[2::])
    cur_state = chartoint(cur_state)
    cur_state = np.array(cur_state)
    
    out = []
    for _ in range(N):
        out.append(cur_state[M-2])
        sum_gen = sum(gen[1::]*cur_state)
        a = sum_gen % 2
        cur_state = np.append(a, cur_state[0:M-2])
    return out
    
        
def chartoint(l):
    out = []
    for i in range(len(l)):
        out.append(int(l[i]))
    return out